rpcs3/Emu/Cell/SPUThread.cpp

```diff
@@ -661,34 +680,142 @@ void SPUThread::do_dma_transfer(const spu_mfc_cmd& args, bool from_mfc)
 	}
 	}
 
-	if (is_get && from_mfc)
+	if (is_get)
 	{
 		//_mm_sfence();
 	}
 }
 
-void SPUThread::process_mfc_cmd()
+bool SPUThread::do_dma_check(const spu_mfc_cmd& args)
 {
-	spu::scheduler::concurrent_execution_watchdog watchdog(*this);
-	LOG_TRACE(SPU, "DMAC: cmd=%s, lsa=0x%x, ea=0x%llx, tag=0x%x, size=0x%x", ch_mfc_cmd.cmd, ch_mfc_cmd.lsa, ch_mfc_cmd.eal, ch_mfc_cmd.tag, ch_mfc_cmd.size);
+	for (u32 i = 0; i < mfc_size; i++)
+	{
+		if ((mfc_queue[i].cmd & ~0xc) == MFC_BARRIER_CMD)
+		{
+			return false;
+		}
 
-	const auto mfc = fxm::check_unlocked<mfc_thread>();
+		if (mfc_queue[i].tag == ch_mfc_cmd.tag)
+		{
+			if (ch_mfc_cmd.cmd & MFC_FENCE_MASK || mfc_queue[i].cmd & MFC_BARRIER_MASK)
+			{
+				return false;
+			}
+		}
+	}
 
-	// Check queue size
-	auto check_queue_size = [&]()
+	return true;
+}
+
+bool SPUThread::do_list_transfer(spu_mfc_cmd& args)
+{
+	vm::reader_lock lock;
+
+	struct list_element
+	{
+		be_t<u16> sb; // Stall-and-Notify bit (0x8000)
+		be_t<u16> ts; // List Transfer Size
+		be_t<u32> ea; // External Address Low
+	} item{};
+
+	while (args.size)
 	{
-		while (mfc_queue.size() >= 16)
+		if (UNLIKELY(item.sb & 0x8000))
 		{
-			if (test(state, cpu_flag::stop + cpu_flag::dbg_global_stop))
+			ch_stall_mask |= (1u << args.tag);
+			ch_stall_stat.push_or(*this, 1u << args.tag);
+			ch_event_stat |= SPU_EVENT_SN;
```

Comment:
this is an old bug, the SN event should be signaled only if the ch_stall_stat count goes from 0 to 1, not everytime a list stalled